server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -422,8 +438,73 @@ public ClusterMetadataManifest getLatestClusterMetadataManifest(String clusterNa
         return fetchRemoteClusterMetadataManifest(clusterName, clusterUUID, latestManifestFileName);
     }
 
+    public String getValidPreviousClusterUUID(String clusterName) throws IOException {
+        Set<String> clusterUUIDs = getAllClusterUUIDs(clusterName);
+        Map<String, ClusterMetadataManifest> latestManifests = getLatestManifestForAllClusterUUIDs(clusterName, clusterUUIDs);
+        List<ClusterMetadataManifest> validChain = createClusterChain(latestManifests);
+        if (validChain.isEmpty()) {
+            return ClusterState.UNKNOWN_UUID;
+        }
+        return validChain.get(0).getClusterUUID();
+    }
+
+    public Set<String> getAllClusterUUIDs(String clusterName) throws IOException {
+        //todo exception handling
+        Map<String, BlobMetadata> clusterUUIDMetadata = clusterUUIDContainer(clusterName).listBlobs();
+        if (clusterUUIDMetadata == null) {
+            return Collections.emptySet();
+        }
+        return Collections.unmodifiableSet(clusterUUIDMetadata.keySet());
+    }
+
+    private Map<String, ClusterMetadataManifest> getLatestManifestForAllClusterUUIDs(String clusterName, Set<String> clusterUUIDs) {
+        Map<String, ClusterMetadataManifest> manifestsByClusterUUID = new HashMap<>();
+        for (String clusterUUID : clusterUUIDs) {
+            try {
+                ClusterMetadataManifest manifest = getLatestClusterMetadataManifest(clusterName, clusterUUID);
+                manifestsByClusterUUID.put(clusterUUID, manifest);
+            } catch (Exception e) {
+                logger.error("Exception in fetching manifest", e);
+                //todo Generic exception should not be handled
+            }
+        }
+        return manifestsByClusterUUID;
+    }
+
+    private List<ClusterMetadataManifest> createClusterChain(Map<String, ClusterMetadataManifest> manifestsByClusterUUID) {
+        Map<ClusterMetadataManifest, String> manifestGraph = manifestsByClusterUUID.values().stream()
+            .collect(Collectors.toMap(Function.identity(), ClusterMetadataManifest::getPreviousClusterUUID));
+        List<ClusterMetadataManifest> validClusterUUIDs = manifestsByClusterUUID.values().stream()
+            .filter(m -> !isGarbageClusterUUID(m) && !manifestGraph.containsValue(m.getClusterUUID())).collect(Collectors.toList());
+        if (validClusterUUIDs.isEmpty()) {
+            logger.info("There is no valid next cluster UUID");
+            return Collections.emptyList();
+        }
+        if (validClusterUUIDs.size() > 0) {
```

Comment:
you mean to check > 1 here

Reply:
We need the error message to be more clear. Valid UUIDs can be multiple but there should only one top level latest UUID and others could be part of the chain in case purging didnt run for some reason.